### PR TITLE
core(implicit parameters and patterns)

### LINKED ISSUE #21: implicit parameter and implicit pattern
### Racket
https://docs.racket-lang.org/syntax-classes/index.html#%28form._%28%28lib._syntax%2Fparse%2Fclass%2Fparen-shape..rkt%29._~7ebraces%29%29

The link is how to match `brace`

### usage

And this is required for `≡` type:
```racket
(data (≡ {A : Type} [a b : A]) : Type
      [refl : (≡ a a)])
```

### Syntax

- [ ] parameter place
    ```racket
    (def (foo {a : A}) : (≡ a a)
      [{a} => refl])
    ```
- [ ] pattern: Using `{}` for implicit bindings and implicit patterns? How to handle destruct in such a pattern? `{suc m}` or `{(suc m)}`?
- [ ] application with implicit arguments

    ```racket
    (def (bar {a : Nat} [b : Nat]) : Unit
      [{a} b => unit])
    (bar z) ; ok, `a` is generated
    (bar z z) ; ok, `a` is `z`, must full provided
    ```

### PR DESCRIPTION
**Pull request by dannypsnl**
resolve #21

### CODE REVIEW

File: k-lib/k/equality.rkt [hunk 1]
```diff
@@ -28,9 +30,9 @@ TODO:
   [A x y refl => refl])
```
Comment (dannypsnl) [inline k-lib/k/equality.rkt:30]: `(≡ y x)` cannot type check now

File: k-lib/k/equality.rkt [hunk 2]
```diff
@@ -28,9 +30,9 @@ TODO:
...
 
 ; Transitivity
-(def (trans {A : Type}
-            [x y z : A]
-            [p : (≡ x y)]
-            [q : (≡ y z)])
-  : (≡ x z)
-  [A x y z refl refl => refl])
+#;(def (trans {A : Type}
```
Comment (dannypsnl) [inline k-lib/k/equality.rkt]: disable for now

File: k-lib/k/equality.rkt [hunk 3]
```diff
@@ -1,10 +1,12 @@
 #lang k/base
 (provide (data-out ≡)
          #;cong
-         symm
-         trans)
+         #;symm
+         #;trans)
 
-(data (≡ [a b : A]) : Type
+(data (≡ {A : Type}
```
Comment (dannypsnl) [inline k-lib/k/equality.rkt:7]: `A` brings locals id-hash for data type

File: k-core/k/def.rkt [hunk 4]
```diff
@@ -65,6 +65,7 @@
          [_ (raise-syntax-error 'bad-pattern
                                 (format "pattern only allows to destruct on constructor")
                                 pat)]))
+     (println (list expr #'ty))
```
Comment (dannypsnl) [inline k-core/k/def.rkt]: debug

**Comment by dannypsnl** [comment]
Idea: normal call on `f` where `f` has implicit parameters is a call with implicit arguments, by generating meta for corresponding implicit parameters. Let **unification** algorithm solve it.

File: k-lib/k/equality.rkt (hunk 3 above)
Comment (dannypsnl) [reply k-lib/k/equality.rkt:7]: no

Comment (dannypsnl) [reply k-lib/k/equality.rkt:7]: the problem is it's a declare level, not pattern level(context of **locals** refer to)

File: k-lib/k/equality.rkt [hunk 5]
```diff
@@ -23,14 +25,14 @@ TODO:
 ; Symmetry
 (def (symm {A : Type}
            [x y : A]
-           [p : (≡ x y)])
-  : (≡ y x)
+           [p : (≡ A x y)])
+  : (≡ A y x)
```
Comment (dannypsnl) [inline k-lib/k/equality.rkt:29]: we cannot, get the type of `A` now

File: k-core/k/def.rkt [hunk 6]
```diff
@@ -65,42 +64,54 @@
...
+       ; application with implicit must provides all implicits
+       ; NOTE: maybe this is not required, but need some researching
+       [(_ p*.full-name ...)
         (define subst-map (make-hash))
-        (check-type #'p*.name (subst #'p*.ty subst-map)
+        (check-type #'p*.full-name (subst #'p*.full-ty subst-map)
                     subst-map)
         ...
-        (with-syntax ([e (stx-map local-expand-expr #'(list p*.name ...))])
+        (with-syntax ([e (stx-map local-expand-expr #'(list p*.full-name ...))])
```
Comment (dannypsnl) [inline k-core/k/def.rkt:75]: where reports `A` has no type

**Review by CYBAI — APPROVED** [review]
:shipit:

**Pull request merged**